src/main/scala/scalaz/stream/async/TaskDriver.scala

```diff
@@ -0,0 +1,164 @@
+   *
+   * Whenever task is completed, following is guaranteed
+   * - if the interrupt was received while evaluating the task (pure code) then `cb` is called with result of the task evaluation
+   * - if the interrupt was received while evaluating the task (async code) then evaluated result is discarded
+   * - otherwise `cb` is called with result of task evaluatin
+   *
+   * @param task    Task to evaluate
+   * @param cb      cb to invoke after evalutation is done, or when task is interrupted
+   */
+  def runTaskAsyncInterruptibly[A](task: Task[A], cb: (Throwable \/ A) => Unit): () => Unit = {
+    sealed trait M
+    case object Interrupt extends M
+    case class Complete(r: Throwable \/ A) extends M
+    case class CompleteBind(r: Any, g: Any => Future[Throwable \/ A]) extends M
+
+    var open: Boolean = true
+
+    def complete(r: (Throwable \/ A)): Unit = {
+      open = false
+      cb(r)
+    }
+
+    def tryCompleteTask(current: Future[Throwable \/ A], actor: => Actor[M]): Unit = {
+      current.step match {
+        case Future.Now(r) => complete(r)
+        case Future.Async(onFinish) =>
+          onFinish((r: Throwable \/ A) => Trampoline.done(actor ! Complete(r)))
+        case Future.BindAsync(onFinish, g) =>
+          onFinish((r: Any) => Trampoline.done(actor ! CompleteBind(r, g)))
+
+      }
+    }
+
+    lazy val actor: Actor[M] = Actor[M]({
+      case Complete(r) if open => complete(r)
+      case CompleteBind(r, g) if open => tryCompleteTask(g(r), actor)
+      case Interrupt if open => complete(left(Task.TaskInterrupted))
+      case _ => //no-op
+
+    })(Strategy.Sequential)
+
+    tryCompleteTask(task.get, actor)
+    () => {actor ! Interrupt}
+  }
+
+
+  /**
+   * Asynchronous execution of this Process. Note that this method is not resource safe unless
+   * callback is called with _left_ side completed. In that case it is guaranteed that all cleanups
+   * has been successfully completed.
+   *
+   * User of this method is responsible for any cleanup actions to be performed by running the
+   * next Process obtained on right side of callback.
+   *
+   * This method returns a function, that when applied, causes the running computation to be interrupted.
+   * That is useful of process contains any asynchronous code, that may be left with incomplete callbacks.
+   * If the evaluation of the process is interrupted, then the interruption is only active if the callback
+   * was not completed before, otherwise interruption is no-op.
+   *
+   * This implementation also guarantees that when `cb` is invoked on right side, then sequence is always non-empty.
+   * This is usefully to run two processes truly in parallel w/o consulting higher level combinator (i.e. wye/merge).
+   *
+   * This implementation as well guaranteed that any onComplete // onFail // onKill code is consulted exactly once
+   * whether the computation was or was not interrupted.
```

Comment:
Here's the core observation that I've been working off of.  Assume the following example:

```scala
await(t) { a => p1 onComplete f1 } onComplete f2
```

Both `p1` and `f1` are a function of `a`, obviously.  However, `f2` is *not* a function of `a` and cannot be!  Now, it is certainly possible to do something terrible like this:

```scala
await(t)(emit) flatMap { a => halt onComplete f3 }
```

In the example above, `f3` may be a function of `a`, but I would argue that the above is a bit of an anti-pattern.  I've never actually seen code which attempts to handle resource cleanup in this fashion, nor would it in fact be sound in general since errors in the outer process (e.g. `await`) would abort the `flatMap` and result in the `onComplete` never getting evaluated.  Which is in fact, intuitive.

Returning to the example at the top of this reply, `f1` is the only finalizer which can *soundly* release `a`.  Thus, unless the user is attempting to do unsound things with the library (such as what `tcp` does), we can rely on them putting their resource cleanup for `t` in `f1`.  Now, invoking `rcv` with `-\/` and appending the continuation will give us the `f2` finalizer (not `a`-dependent), which we run and satisfy our "exactly once" guarantee.  Invoking it *again* with `\/-` but *not* appending the continuation will give us `f1`, which is exactly and only the finalizer that releases `a`.

You'll note that, in your example above, `cleanup` is not parameterized on `a`!  If it were, then it would be clear just from the types that it could not possibly be invoked on interrupt, and can only be run *after* completion.  Cleanup that is not parameterized on `a`, either via function parameters or closure, is much less interesting.